json.go

```diff
@@ -0,0 +1,797 @@
+			jobj := val.(JSONObject)
+			val, err = jobj.GetValue()
+		} else if j.dpiJsonNode.oracleTypeNum == C.DPI_ORACLE_TYPE_JSON_ARRAY {
+			jarr := val.(JSONArray)
+			val, err = jarr.GetValue()
+		} else {
+			err = nil
+		}
+	}
+	return
+}
+
+func (j JSONScalar) getOracleJSONType() C.dpiOracleTypeNum {
+	return j.dpiJsonNode.oracleTypeNum
+}
+
+func (js *JSONScalar) Close() error {
+	if js.isMemOwned {
+		C.godror_dpiJsonfreeMem(js.dpiJsonNode)
+	}
+	return nil
+}
+
+// JSON holds the JSON data to/from Oracle.
+// It is like a root node in JSON tree.
+type JSON struct {
+	dpiJson       *C.dpiJson
+	isMemOwned    bool // true indicates associated C memory needs to be freed.
+	oracleTypeNum C.dpiOracleTypeNum
+}
+
+// During Fetch the JSON struct is created.
+// Explicit creation is not supported.
+func createJSON(js *C.dpiJson, mutable bool) JSON {
+	ntype := C.dpiOracleTypeNum(js.topNode.oracleTypeNum)
+	json := JSON{dpiJson: js, isMemOwned: mutable, oracleTypeNum: ntype}
+	return json
+}
+
+func (j JSON) getOracleJSONType() C.dpiOracleTypeNum {
+	return j.oracleTypeNum
+}
+
+func (j JSON) Get(data *Data, opts JSONOption) error {
+	var node *C.dpiJsonNode
+	if C.dpiJson_getValue(j.dpiJson, C.uint32_t(opts), (**C.dpiJsonNode)(unsafe.Pointer(&node))) == C.DPI_FAILURE {
+		return ErrInvalidJSON
+	}
+	jsonNodeToData(data, node)
+	return nil
+}
+
+// Returns JSONObject from JSON
+func (j JSON) GetJSONObject(jsobj *JSONObject, opts JSONOption) error {
+	var node *C.dpiJsonNode
+	var d Data
+	if C.dpiJson_getValue(j.dpiJson, C.uint32_t(opts), (**C.dpiJsonNode)(unsafe.Pointer(&node))) == C.DPI_FAILURE {
+		return ErrInvalidJSON
+	}
+	jsonNodeToData(&d, node)
+	if C.dpiOracleTypeNum(node.oracleTypeNum) != C.DPI_ORACLE_TYPE_JSON_OBJECT {
+		return ErrInvalidType
+	}
+
+	*jsobj = JSONObject{dpiJsonNode: node,
+		dpiJsonObject: C.dpiData_getJsonObject(&(d.dpiData)),
+		isMemOwned:    j.isMemOwned}
+	return nil
+}
+
+// Returns JSONArray from JSON
+func (j JSON) GetJSONArray(jsarr *JSONArray, opts JSONOption) error {
+	var node *C.dpiJsonNode
+	if C.dpiJson_getValue(j.dpiJson, C.uint32_t(opts), (**C.dpiJsonNode)(unsafe.Pointer(&node))) == C.DPI_FAILURE {
+		return ErrInvalidJSON
+	}
+	var d Data
+	jsonNodeToData(&d, node)
+	if C.dpiOracleTypeNum(node.oracleTypeNum) != C.DPI_ORACLE_TYPE_JSON_ARRAY {
+		return ErrInvalidType
+	}
+	*jsarr = JSONArray{dpiJsonNode: node,
+		dpiJsonArray: C.dpiData_getJsonArray(&(d.dpiData)),
+		isMemOwned:   j.isMemOwned}
+	return nil
+}
+
+// Returns JSONScalar from JSON
+func (j JSON) GetJSONScalar(js *JSONScalar, opts JSONOption) error {
+	var node *C.dpiJsonNode
+	if C.dpiJson_getValue(j.dpiJson, C.uint32_t(opts), (**C.dpiJsonNode)(unsafe.Pointer(&node))) == C.DPI_FAILURE {
+		return ErrInvalidJSON
+	}
+	*js = JSONScalar{dpiJsonNode: node,
+		isMemOwned: j.isMemOwned}
+	return nil
+}
+
+// Returns a Go type Value from JSON
+func (j JSON) GetValue(opts JSONOption, val *interface{}) (err error) {
```

Comment:
Why not return the `(val interface{}, err error)` ?

Reply:
I was trying to make it look similar to json Unmarshal. is it fine? I can change it ...
func Unmarshal(data []byte, v interface{}) error

Reply:
Let me check, you mean to remain consistent with other type; JSONObject, . Let me validate on these.

Reply:
Unmarshal uses the destination interface{} (NOT *interface{} !) - which must contain a pointer -
to deduce the wanted type of data. It could Unmarshal a js number (float64) into a string, int, godror.Number etc.

Here GetValue does nothing like, it decides the returned data type by opts.

Reply:
Made to return pointers in all the API.